x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
     private static boolean isInternalAction(String action) {
         return INTERNAL_PREDICATE.test(action);
     }
+
+    public static boolean isParentAction(String action) {
+        return PRE_AUTHORIZED_CHILD_ACTIONS_BY_PARENT.containsKey(action);
+    }
+
+    public static boolean shouldPreAuthorizeChildAction(final String parent, final String child) {
+        final Set<String> children = PRE_AUTHORIZED_CHILD_ACTIONS_BY_PARENT.get(parent);
+        return children != null && (parent.equals(child) || children.contains(child));
+    }
+
+    public static void putParentAuthorizationContext(ThreadContext threadContext, AuthorizationEngine.AuthorizationContext context) {
+        AuthorizationEngine.AuthorizationContext existing = threadContext.getTransient(
+            AuthorizationServiceField.PARENT_AUTHORIZATION_CONTEXT_KEY
+        );
+        if (existing == null) {
+            threadContext.putTransient(AuthorizationServiceField.PARENT_AUTHORIZATION_CONTEXT_KEY, context);
+        } else if (existing.getAction().equals(context.getAction()) == false) {
+            throw new AssertionError("parent authorization context already set in thread context");
+        }
+    }
```

Comment:
Nit, please create a class for `PreAuthorization` don't dump them in the `AuthorizationUtils` class.

Reply:
I introduced new [PreAuthorizationUtil](https://github.com/elastic/elasticsearch/blob/bed1362aeac80b4988b3c1f10796a9e0803ce838/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java) which defines all methods for handling pre-authorizations.